src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlySpanND.cs

```diff
@@ -0,0 +1,468 @@
+            _strides = SpanNDHelpers.CalculateStrides(Rank, lengths);
+        }
+
+        /// <summary>
+        /// Creates a new read-only span over the target unmanaged buffer.  Clearly this
+        /// is quite dangerous, because we are creating arbitrarily typed T's
+        /// out of a void*-typed block of memory.  And the length is not checked.
+        /// But if this creation is correct, then all subsequent uses are correct.
+        /// </summary>
+        /// <param name="pointer">An unmanaged pointer to memory.</param>
+        /// <param name="lengths">The <see cref="System.ReadOnlySpan{T}"/> with the lengths of each dimension.</param>
+        /// <param name="isPinned">If the underlying data is pinned.</param>
+        /// <param name="strides">The <see cref="System.ReadOnlySpan{T}"/> with the lengths of each stride.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when <typeparamref name="T"/> is reference type or contains pointers and hence cannot be stored in unmanaged memory.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when the specified length is negative.
+        /// </exception>
+        [CLSCompliant(false)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe ReadOnlySpanND(void* pointer, ReadOnlySpan<nint> lengths, bool isPinned, ReadOnlySpan<nint> strides = default)
+        {
+            _linearLength = SpanNDHelpers.CalculateTotalLength(lengths);
+
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
+            if (_linearLength < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+            if (_linearLength != SpanNDHelpers.CalculateTotalLength(lengths))
+                ThrowHelper.ThrowArgument_LengthsMustEqualArrayLength();
+
+            _isPinned = isPinned;
+
+#pragma warning disable CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
+            _reference = ref *(T*)pointer;
+#pragma warning restore CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
+
+            _lengths = lengths.ToArray();
+            _strides = strides.ToArray();
+            if (strides == ReadOnlySpan<nint>.Empty)
+                _strides = SpanNDHelpers.CalculateStrides(Rank, lengths);
+        }
+
+        // Constructor for internal use only. It is not safe to expose publicly, and is instead exposed via the unsafe MemoryMarshal.CreateReadOnlySpan.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ReadOnlySpanND(ref T reference, ReadOnlySpan<nint> lengths, ReadOnlySpan<nint> strides, bool isPinned)
+        {
+            _linearLength = SpanNDHelpers.CalculateTotalLength(lengths);
+            Debug.Assert(_linearLength >= 0);
```

Comment:
If this is asserting, then I would've expected we already computed the `linearLength` and it could be passed in instead

Reply:
Fixed/removed.